### PR TITLE
detailed 'cdm_enum_fk_candidates()'

### LINKED ISSUE #24: Explain in more detail why column is not a foreign key
Mention which values are not contained in the referred primary key. In #17 I already added printing of duplicate values for primary key candidates.

### PR DESCRIPTION
**Pull request by TSchiefer**
- checks now for class compatibility
- if compatible, checks if subset
- lists in column `why`:
    1. nothing if candidate
    1. class `...` doesn't match pk-col class `...`
    1. following values are not in pk-col: ...

fixes #24

### CODE REVIEW

File: R/foreign-keys.R [hunk 1]
```diff
@@ -189,20 +192,54 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
...
+  # which columns contain subset of values of PK-col?
+  col_candidates <- cols_class_match[map_lgl(
+    cols_class_match,
+    ~ is_subset(tbl, !!.x, ref_tbl, !!ref_tbl_pk))]
+  tibble_candidates <- tibble(column = col_candidates,
+    candidate = TRUE,
+    why = "")
+
+  # which columns of same class contain additional values and which ones?
+  set_of_pk_vals <- pull(arrange(distinct(ref_tbl, !!ensym(ref_tbl_pk))))
```
Comment (krlmlr) [inline R/foreign-keys.R]: Can you please use the following idiom to compute mismatches:

```r
tryCatch(
  anti_join(...) %>%
     utils::head(MAX_COMMAS + 1) %>%
     collect(),
  error = identity
)
```

Then, check for `is_condition(.)` and, if yes, use `conditionMessage(.)` as reason.

File: tests/testthat/helper-src.R [hunk 2]
```diff
@@ -8,10 +8,10 @@ message("for ccheck_cardinality...()")
 d1 <- tibble::tibble(a = 1:5, b = letters[1:5])
 d2 <- tibble::tibble(a = c(1, 3:6), b = letters[1:5])
 d3 <- tibble::tibble(c = 1:5)
-d4 <- tibble::tibble(c = c(1:5, 5))
+d4 <- tibble::tibble(c = c(1:5, 5L))
 d5 <- tibble::tibble(a = 1:5)
 d6 <- tibble::tibble(c = 1:4)
-d7 <- tibble::tibble(c = c(1:5, 5, 6))
+d7 <- tibble::tibble(c = c(1:5, 5L, 6L))
```
Comment (krlmlr) [inline tests/testthat/helper-src.R]: Can you tweak the test so that we check integer/numeric compatibility?

Comment (TSchiefer) [reply tests/testthat/helper-src.R]: Just so I got it right: we want a test that compares numeric columns with integer columns. I just checked, what `anti_join()` does locally:
``` r
suppressPackageStartupMessages(library(dplyr))
library(tibble)
t1 <- tibble(a = as.numeric(1:5))
t1
#> # A tibble: 5 x 1
#>       a
#>   <dbl>
#> 1     1
#> 2     2
#> 3     3
#> 4     4
#> 5     5
t2 <- tibble(a = 1:5)
t2
#> # A tibble: 5 x 1
#>       a
#>   <int>
#> 1     1
#> 2     2
#> 3     3
#> 4     4
#> 5     5
anti_join(t1, t2, by = "a")
#> # A tibble: 0 x 1
#> # … with 1 variable: a <dbl>
```

<sup>Created on 2019-09-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
But on DB there might/will be differences. So is the test supposed to show the differences between DB- and local behaviour?

Comment (krlmlr) [reply tests/testthat/helper-src.R]: I'd expect this behavior for all databases.

File: R/foreign-keys.R [hunk 3]
```diff
@@ -152,31 +152,34 @@ cdm_rm_fk <- function(dm, table, column, ref_table) {
...
 #' @param ref_table A table with a primary key.
 #'
-#' @description Which columns are foreign candidates of a table, referencing the primary key column of another [`dm`] object's table?
-#' `cdm_enum_fk_candidates()` checks first, if `ref_table` has a primary key set. Then it determines
-#' for each column of `table`, if this column contains only a subset of values of the primary key column of
-#' `ref_table` and is therefore a candidate for a foreign key from `table` to `ref_table`.
+#' @details `cdm_enum_fk_candidates()` checks first, if `ref_table` has a primary key set. Then
+#' it determines for each column of `table`, if it has the same class as the primary key of `ref_table`.
+#' If this is the case a check is performed, if the column contains only a subset of values of the
+#' primary key column of `ref_table`. If this is `TRUE`, this column is a candidate for a foreign key
```
Comment (krlmlr) [inline R/foreign-keys.R]: Can you use bullets or an enumeration to describe the rules?

```suggestion
#' primary key column of `ref_table`.
#' This check uses [anti_join()] which implicitly checks type compatibility.
#' If this is `TRUE`, this column is a candidate for a foreign key
```

File: R/foreign-keys.R [hunk 4]
```diff
@@ -185,23 +188,32 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
...
+why <- function(dm, t1, colname, t2, t2_name, pk) {
+  names(pk) <- colname
+  test <- tryCatch(anti_join(
+    select(t1, !!sym(colname)), select(t2, !!sym(pk)), by = pk) %>%
+      utils::head(MAX_COMMAS + 1) %>%
+      pull(), error = identity)
+  if (is_condition(test)) {
+    return(conditionMessage(test))
+  }
+  if (is_empty(test)) return("") else {
```
Comment (krlmlr) [inline R/foreign-keys.R]: We don't need the `else` here.

File: R/foreign-keys.R [hunk 5]
```diff
@@ -185,23 +188,32 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
...
+    why = map_chr(column, ~why(dm, tbl, .x, ref_tbl, ref_table_name, ref_tbl_pk))
+  ) %>%
+    mutate(candidate = ifelse(why == "", TRUE, FALSE)) %>%
+    select(column, candidate, why) %>%
+    mutate(arrange_col = str_sub(why, 1, 3)) %>%
+    arrange(desc(candidate), desc(arrange_col), column) %>%
+    select(-arrange_col)
 })
+
+why <- function(dm, t1, colname, t2, t2_name, pk) {
```
Comment (krlmlr) [inline R/foreign-keys.R]:
```suggestion
check_fk <- function(dm, t1, colname, t2, t2_name, pk) {
```

File: tests/testthat/test-foreign-key-functions.R [hunk 6]
```diff
@@ -159,23 +159,89 @@ test_that("cdm_rm_fk() works as intended?", {
...
+    "minute",         FALSE,      "Can't join on 'minute' x 'faa' because of incompatible types (numeric / character)",
+    "month",          FALSE,      "Can't join on 'month' x 'faa' because of incompatible types (integer / character)",
+    "sched_arr_time", FALSE,      "Can't join on 'sched_arr_time' x 'faa' because of incompatible types (integer / character)",
+    "sched_dep_time", FALSE,      "Can't join on 'sched_dep_time' x 'faa' because of incompatible types (integer / character)",
+    "year",           FALSE,      "Can't join on 'year' x 'faa' because of incompatible types (integer / character)",
+    "time_hour",      FALSE,      "cannot join a POSIXct object with an object that is not a POSIXct object"
+  )
+
+  expect_identical(
+    cdm_enum_fk_candidates(cdm_nycflights13(), flights, airports),
```
Comment (krlmlr) [inline tests/testthat/test-foreign-key-functions.R]:
```suggestion
    cdm_enum_fk_candidates(cdm_nycflights13(), flights, airports) %>%
      mutate(why = if_else(why != "", "<reason>", ""),
```

File: R/foreign-keys.R [hunk 7]
```diff
@@ -185,23 +188,32 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
...
+  names(pk) <- colname
+  test <- tryCatch(anti_join(
+    select(t1, !!sym(colname)), select(t2, !!sym(pk)), by = pk) %>%
```
Comment (krlmlr) [inline R/foreign-keys.R]: ~~We need to remove dupes on the server: anti_join %>% distinct %>% arrange %>% head %>% collect %>% pull~~

We also might want the total count and the mismatch percentage to detect weak relationships:

```r
t1_join <- t1 %>% select(value = !!sym(colname))
t2_join <- t2 %>% select(value = !!sym(pk)) %>% mutate(match = 1L)

left_join(t1_join, t2_join, by = "value") %>%
  mutate(mismatch_or_null = if_else(is.na(match), value, NULL)) %>%
  count(mismatch_or_null) %>%
  ... %>%
  collect()
```

Example result for `flights$dest`:

1334 rows (8.5%) of the values not in `airports$faa`: SJU (280), BQN (110), 3 more, …

File: R/foreign-keys.R [hunk 8]
```diff
@@ -185,23 +194,55 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
...
+check_fk <- function(dm, t1, t1_name, colname, rows_t1, t2, t2_name, pk) {
+
+  t1_join <- t1 %>% select(value = !!sym(colname))
+  t2_join <- t2 %>% select(value = !!sym(pk)) %>% mutate(match = 1L)
+
+  res_tbl <- tryCatch(
+    left_join(t1_join, t2_join, by = "value") %>%
+      mutate(mismatch_or_null = if_else(is.na(match), value, NULL)) %>%
+      count(mismatch_or_null) %>%
+      filter(!is.na(mismatch_or_null)) %>%
```
Comment (krlmlr) [inline R/foreign-keys.R]: This only removes one row, and I think we need it to compute percentages?

Comment (krlmlr) [reply R/foreign-keys.R]:
```r
left_join(t1_join, t2_join, by = "value") %>%
  # if value is NULL, this also counts as a match -- consistent with fk semantics
  mutate(mismatch_or_null = if_else(is.na(match), value, NULL)) %>%
  count(mismatch_or_null) %>%
  ungroup() %>% # dbplyr problem?
  mutate(n_mismatch = sum(if_else(is.na(value, 0, 1), na.rm = TRUE))) %>%
  mutate(n_total = sum(n, na.rm = TRUE)) %>%
  arrange(desc(n)) %>%
  head(MAX_COMMAS + 1L) %>%
  collect()
```

**Pull request merged**